framework/include/materials/MaterialPropertyInterface.h

```diff
@@ -369,6 +388,9 @@ class MaterialPropertyInterface
 
   /// Storage for the boundary ids created by BoundaryRestrictable
   const std::set<BoundaryID> & _mi_boundary_ids;
+
+  /// address stable pointer storage to enable returning optional property pointers via reference
+  std::list<const PropertyValue *> _optional_material_pointer_storage;
```

Comment:
Why not a vector of unique_ptrs?

Reply:
Because I need to return a reference to a pointer and the address stable list is the easiest way to do that. No need to use smart pointers here.